tensorflow/core/kernels/fused_embedding/embedding_lookup_sparse_post_op.cc

```diff
@@ -0,0 +1,450 @@
+
+namespace {
+  inline int64 partitioned_indices(std::vector<std::tuple<size_t, const int64*>> &indices, int indice_dim, int64 id) {
+      int indices_num = indices.size();
+      int64 rows = 0;
+      for (int i = 0; i < indices_num; ++i) {
+          size_t sub_nnz = std::get<0>(indices[i]);
+          rows += sub_nnz;
+          if (rows > id) {
+              int idx = id - (rows - sub_nnz);
+              return std::get<1>(indices[i])[idx * indice_dim];
+          }
+      }
+  }
+
+  inline const float* const partitioned_embedding_tables(std::vector<std::tuple<size_t, const float*>> &embedding_tables,
+                                                          int embedding_size, int64 id) {
+      int tables_num = embedding_tables.size();
+      int64 rows = 0;
+      for (int i = 0; i < tables_num; ++i) {
+          size_t sub_nnz = std::get<0>(embedding_tables[i]);
+          rows += sub_nnz;
+          if (rows > id) {
+              int idx = id - (rows - sub_nnz);
+              return &(std::get<1>(embedding_tables[i])[idx * embedding_size]);
+          }
+      }
+  }
+
+  static void sparse_partitioned_gather(int64 input_size,
+                          std::vector<std::tuple<size_t, const int64*>> &indices, int indice_dim, int rows,
+                          std::vector<std::tuple<size_t, const float*>> &embedding_tables, float *output,
+                          const int64_t embedding_size, SparseSegmentReductionOperation operation,
+                          const bool set_empty_row_zero, const int *empty_row) {
+    // Record how many values in each row
+    uint64_t* row_values = new uint64_t[rows];
+    memset(row_values, 0, rows * sizeof(uint64_t));
+    // output_buffer is output buffer
+    float* output_buffer = new float[rows*embedding_size];
+    memset(output_buffer, 0, rows*embedding_size * sizeof(float));
+
+#ifdef __AVX512F__
+    auto avx512_add = [](const float* input, uint64_t input_idx,
+                          float* output, uint64_t output_idx, const int64_t num) {
+      constexpr size_t float_displacement = 4;
+      constexpr size_t float_alignment = 16;
+      int64_t quotient = num >> float_displacement;
+      int64_t remainder = num & 0x000F;
+
+      for (int64_t j = 0; j < quotient; ++j) {
+        int64_t offset = j << float_displacement;
+        __m512 a = _mm512_loadu_ps(&input[input_idx + offset]);
+        __m512 b = _mm512_loadu_ps(&output[output_idx + offset]);
+        a = _mm512_add_ps(a, b);
+        _mm512_storeu_ps(&output[output_idx + offset], a);
+      }
+
+      if (remainder != 0) {
+        __mmask16 mask = 0xffff >> (float_alignment - remainder);
+        int64_t offset = quotient << float_displacement;
+        __m512 a = _mm512_mask_loadu_ps(_mm512_setzero_ps(), mask, &input[input_idx + offset]);
+        __m512 b = _mm512_mask_loadu_ps(_mm512_setzero_ps(), mask, &output[output_idx + offset]);
+        a = _mm512_mask_add_ps(_mm512_setzero_ps(), mask, a, b);
```

Comment:
above lines depends on compiler for the optimization, most compilers may already do the optimization, but code like below could be better:
__m512 zero = _mm512_setzero_ps();
__m512 a = _mm512_mask_loadu_ps(zero, mask, &input[input_idx + offset]);
__m512 b = _mm512_mask_loadu_ps(zero, mask, &output[output_idx + offset]);
a = _mm512_mask_add_ps(zero, mask, a, b);